fem/nonlinearform.cpp

```diff
@@ -109,13 +134,25 @@ const Vector &NonlinearForm::Prolongate(const Vector &x) const
 
 void NonlinearForm::Mult(const Vector &x, Vector &y) const
 {
+   const Vector &px = Prolongate(x);
+   if (P) { aux2.SetSize(P->Height()); }
+
+   // If we are in parallel, ParNonLinearForm::Mult uses the aux2 vector.
+   // In serial, place the result directly in y.
+   Vector &py = P ? aux2 : y;
+
+   if (ext)
+   {
+      ext->Mult(px, py);
+      py.HostRead();
```

Comment:
This `HostRead` seems out of place.

The principle we try to use when moving data to/from GPU is to do it right before a kernel is executed, based on where the kernel needs the data.

This way the kernel (currently implemented for CPU only, I assume, because of the `HostRead`) will always have the data where it needs it, regardless of which code path lead to it; also if it is ported later to GPU, it will be easy to update the `HostRead` with a `Read` because the code is in the same place.

Reply:
I forgot to address the GPU related stuff here. Removed line 147.